Fix Devmapper issue: Power off the VM while loading a image, couldn't load it after the VM bootup

moypray opened the pull request:
Issue Description:
    While running loading image test, power off or restart the VM, and then there are
    some chances that we can not load the image. And "Error running
    deviceCreate (createSnapDevice) dm_task_run failed" will be reported.

    Reproduce Steps:
    > 1. run  `docker load -i xxx.tar`
    > 2. virsh restart VM; # restart the VM.
    > 3. After startup, run `docker load -i xxx.tar`, will fail to import the image

    Analysis:
    From syslog, we found that docker was executing "Umount Device" then VM powered restart. And found
    two failure reasons:
    > 1. Rollback operation only remove the device on DM thin pool, not remove the device in memory
    > 2. TransactionData or metadata not flushed to Disk.

    Solution:
    > 1. Rollback operation for DM, should remove the devices cache in DM driver.
    > 2. When restore layers and images, check whether the device exists in graphdriver.
         If desen't, remove the layer( and the metadata) and do not load the image.

    Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Fix a devicemapper issue:
 when loading a image, restart the VM, when the system bootup, there are some chances to fail to load the image again.

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

coolljt0725 commented:
ping @rhvgoyal

moypray commented:
ping @coolljt0725  @rhvgoyal @vdemeester  I found two issues in dm driver: 
> 1. Rollback operation only remove the device from DM thin pool, doesn't remove the device in memory
> 2. TransactionData or metadata not flushed to Disk.

For  the first one, we can remove the cache from `devices.Devices` of the device. 
For the transaction issue, I have added a check while loading the layer metadata, if the device doesn't exist, Remove the layer and layer metadata; And further, if a layer doesn't exist whiling loading the image, just do not load it to memory, skip it.

coolljt0725 commented:
ping @tonistiigi   @aaronlehmann

moypray commented:
Hi  I have done some investigation about this issue:
Reproduce Steps:
> 1.  Running `docker load -i  ubuntu.tar`
> 2.  Do NOT wait the load process done, `virsh reboot VM`
> 3.  After VM reboot,  run `docker load -i ubuntu.tar`

I got an error as below:
`[root@localhost ~]# docker load -i ubuntu.tar  
devicemapper: Error running deviceCreate (createSnapDevice) dm_task_run failed`

From docker startup log I found that the device is rollback by devmapper driver:
`DEBU[0001] devmapper: Rolling back open transaction:TransactionID=249, TransactionID=250 hash=4d39d33fd6216f5a505b87380b81dcc6ca0c36cad55ef91eee6c35432494c5c8 device_id=157`

Rollback operation of the device, Devmapper driver will remove the device, but **DOSE NOT remove the data in memory, and layer data.** 

When we load the image, docker think the device exists on devmapper, but actually only the metadata exists and device not.

Logs:
`Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.326116302-04:00" level=debug msg="devmapper: AddDevice(hash=82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081 basehash=286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a)"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.326224693-04:00" level=debug msg="devmapper: Failed to read /var/lib/docker/devicemapper/metadata/82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081 with err: open /var/lib/docker/devicemapper/metadata/82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081: no such file or directory"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.393903229-04:00" level=debug msg="devmapper: Error creating snap device: devicemapper: Error running deviceCreate (createSnapDevice) dm_task_run failed"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.393936257-04:00" level=debug msg="devmapper: AddDevice(hash=82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081 basehash=286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a) END"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.393998904-04:00" level=debug msg="devmapper: DeleteDevice(hash=286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a syncDelete=false) START"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.394014969-04:00" level=debug msg="devmapper: deactivateDevice(286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a)"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.394070839-04:00" level=debug msg="devmapper: deactivateDevice END(286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a)"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.400993303-04:00" level=debug msg="devmapper: Error deleting device: devicemapper: Error running DeleteDevice dm_task_run failed"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.403022184-04:00" level=debug msg="devmapper: DeleteDevice(hash=286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a syncDelete=false) END"
Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.403085295-04:00" level=error msg="Error releasing layer sha256:e7ebc6e16708285bee3917ae12bf8d172ee0d7684a7830751ab9a1c070e7a125: devicemapper: Error running DeleteDevice dm_task_run failed"
`

This happens not each time. But if there are lots of devices in DM, but not records in DM MetaData. It will happens more frequently. So I think the transaction data or meta data of the this device was not sync while reboot the system. 

So I fix the following issue:
> 1. Write devmapper meta data using sync mode 
> 2. When rollback happens, remove the device, and remove the devices data in docker memory.
> 3. When loading the Layer data, check if the graphdriver has the cache-ID data. if not, remove the layer and layer metadata from disk and memory.
> 4.  When loading image, if some or one layer is broken, ignore it , continue to bootup. the image will not be loaded. User should pull it again.

aaronlehmann commented on layer/layer_store.go:
```diff
@@ -106,11 +106,17 @@ func NewStoreFromGraphDriver(store MetadataStore, driver graphdriver.Driver) (St
+		if err != nil {
+			ls.store.Remove(layer)
+		}
+	}()
```
Is removing the metadata really appropriate if the layer couldn't be loaded? I could see this being destructive if a temporary condition causes failures (for example, something is not mounted).

cc @dmcgowan

coolljt0725 replied:
The main problem to keep the metadata on failures is we can't load/pull the same image any more, it will cause rename error something like 
```
rename /var/lib/docker/image/devicemapper/layerdb/tmp/layer-870197081 /var/lib/docker/image/devicemapper/layerdb/sha256/b257bb4b4aeaa5590b31c125581c301d69352345c09fe860a35de4c7cf212ae7: directory not empty
```
we have to remove the directory manually 

 @aaronlehmann  Instead of removing the metadata in a `defer`, how about remove it on failuer which is can't caused by temporay condition? And which failure will caused by temporary condition?

moypray replied:
`loadLayer` loads the layer metadata from disk, if failure means something wrong with the metadata. I think if we do not remove the data on disk , could cause the layer data in memory and disk do not keep the same. And also cause the  issue as collljt0725 mentioned.

cpuguy83 replied:
ping @dmcgowan

moypray committed:
Fix Devmapper issue: Power off the VM while loading a image, couldn't load it after the VM bootup

Issue Description:
While running loading image test, power off or restart the VM, and then
there are
some chances that we can not load the image. And "Error running
deviceCreate (createSnapDevice) dm_task_run failed" will be reported.

Reproduce Steps:
> 1. run  `docker load -i xxx.tar`
> 2. virsh restart VM; # restart the VM.
> 3. After startup, run `docker load -i xxx.tar`, will fail to import
> the image

Analysis:
From syslog, we found that docker was executing "Umount Device" then VM
powered restart. And found
two failure reasons:
> 1. Rollback operation only remove the device on DM thin pool, not
> remove the device in memory
> 2. TransactionData or metadata not flushed to Disk.

Solution:
> 1. Rollback operation for DM, should remove the devices cache in DM
> driver.
> 2. When restore layers and images, check whether the device exists in
> graphdriver.
     If desen't, remove the layer( and the metadata) and do not load the
image.

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

rhvgoyal commented:
I don't think problem is in devicemapper layer. It removes device id as well as metadata file if transaction was not complete. And memory is not an issue at all as you rebooted the VM.

I think problem is that layer code does not implement the notion of transaction. So it might think that an image got created in devicemapper while it was not because VM was rebooted. So upper layers first should wait for a success message from Create() call and then only mark image data valid. This is evident from following message.

AddDevice(hash=82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081 basehash=286ade1d79c9a7eb79a796ef3afb536d6e18e71c1ae2e3d6b4a9f42d641d1a1a)" Mar 23 17:05:26 localhost.localdomain docker[1266]: time="2017-03-23T17:05:26.326224693-04:00" level=debug msg="devmapper: Failed to read /var/lib/docker/devicemapper/metadata/82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081 with err: open /var/lib/docker/devicemapper/metadata/82ab85c16d5ed599292f202890a6fed7899b99bf2f91654321b1c638f2eac081: no such file or directory"

There is no metadata file belonging to id `82ab85c16d`. I suspect that this is the file/device which got deleted during rollback and upper layers are assuming that creation of this layer succeeded during crash.

moypray commented:
@rhvgoyal 
The rollback operation of transaction is to remove the device and metadata of the device. while an open transaction exists, and during next startup, the device and metadata will be removed. See function `rollbackTransaction`. (just like my case)
So the metadata file does not exists.


Yes, my fix contains checking layer exists or not while restore layer metadata.  But the issue is relative with DM transaction.  Surely the error is not relative the the memory in `device.Devices` map, but remove it  should be a better choice.

Of course, if the layer does not exist, and remove the metadata, we should consider the image too.

rhvgoyal commented:
@moypray can you describe what's the problem? 

There is no need to query if graph driver has layer or not and then remove it.

Layer data should be marked valid, only when "create" call has been successful. 

That way, if create was not successful due to power off, DM will rollback transaction and layer logic can see that a layer's data is not valid and then delete that data. 

Querying graph driver and removing maks if there is a real bug in devicemapper or any of the graph driver and they lost image metadata which they should not have.

moypray commented:
Hi @rhvgoyal  sorry for late reply,  I am so busy these days....

As DM rollback transaction was operated, the device on DM thin pool will be removed. But the LayerDB metadata (`layer/layer_store.go  function restore()`) is still on disk(not the DM metadata), when docker starts up, will load the layer metadata from disk, and we need to remove the layerDB from layer store and disk.  Or we will failed to load/pull a image which contains this layer, as the upper logic of pull/load just check if the metadata exists on layer store.

```
+func (ls *layerStore) loadLayer(layer ChainID) (l *roLayer, err error) {
  	cl, ok := ls.layerMap[layer]		  	cl, ok := ls.layerMap[layer]
  	if ok {		  	if ok {
  		return cl, nil		  		return cl, nil
  	}		  	}
 +	defer func() {
 +		// If failed to load the layer, remove the layer metadata.
 +		if err != nil {
 +			ls.store.Remove(layer)
 +		}
 +	}()
```

So we need to check if the devices exist on graphdriver level  or not when docker starts up.

rhvgoyal commented:
Instead of checking if layer exist or not, we should not create layer data or mark it valid only after create() has succeeded from graph drivers.

If layer does not exist at graph driver level, that's a bug. So checking it after that fact is not the right solution.

moypray commented:
> If layer does not exist at graph driver level, that's a bug. So checking it after that fact is not the right solution.  

You are right, I agree with you, but I tried to fix it in devmapper. but it is quite difficult to fix it. The scenario is "dm is operated in Driver", we restart the system. the device in thin pool is damaged, we must delete it after system bootup. and can not recover it.
Do you have any idea about this?


And this condition will only appear on docker startup process. `restore` function is only called while the daemon setup. it is a workaround but could help.

rhvgoyal commented:
Fix is not in dm. Device mapper layer is just fine. Problem is generic layer logic which is caller of Create(). I think we should do following.

- Create layer data with a bit/field marked as "in progress"
- Call devicemapper graph driver Create()
- Upon successful call mark that field as "created".

Now after a crash if layer logic finds a layer/mounted-layer data which status "in progress", you know that we crashed and operation might not have completed. In that case, check if layer exists() at graph driver level and if does exist, delete it.

Otherwise cleanup that layer in layer store/mounted layer.

IOW, this will be equivalent of transaction mechanism where you are rolling back layer/mounted-layer changes if operation did not complete and crash happened in between.

tonistiigi commented:
@rhvgoyal Looking at https://github.com/docker/docker/blob/1a8206ca1fefdcc50577248bf27107499fa3d55e/layer/layer_store.go#L292 we don't write anything in layerstore before `driver.Create` has returned.

rhvgoyal commented:
@tonistiigi Right, looks like layer data is not being saved till Create() is complete. I think something is missing. The very fact rollback took place, that means Create() did not complete. And There are calls to AddDevice()  upon restart where base device is missing. That means higher layers think that base device exists while this might be the one which got rolled back.

I think I need more information. 

@moypray What's the docker version you are using. Also can you run it again and get me all the docker logs. (before crash and after crash).

moypray commented:
Hi @rhvgoyal , I am using 1.11.2 version.  But the logic of `Register` function is almost the same.

I have cherry-pick some PR from master branch. such as the sync operation of writing metadata(layer,image etc.).

Perhaps we just need the sync operation to write DM metadata? I am confused now.
https://github.com/docker/docker/pull/31934/files#diff-4c195862cdadcaf026de87c0f6d7c313R333

I have tried to patch the sync operation of DM metadata only, but it doesn't work.

moypray commented:
@rhvgoyal  @tonistiigi , thanks for your input, I found some logs: 

Maybe we just need the sync operation on metadata for DM, but I am not sure if this could solve all the problem.

Before system shutdown, docker was operating on a layer/device 
(0f1759dafc79227fabed4d2f2af8aa12963bdd7eae8dc881fc5063edad06925d):  
```
 29438 Mar 15 11:16:07 V2R1C00B013-GUESTOS-FS-KVM-X64 docker: time="2017-03-15T11:16:07.275107074Z" level=debug msg="Untar time: 5.630688415s"
  29439 Mar 15 11:16:07 V2R1C00B013-GUESTOS-FS-KVM-X64 docker: time="2017-03-15T11:16:07.275163687Z" level=debug msg="devmapper: UnmountDevice(hash=0f1759dafc79227fabed4d2f2af8aa12963bdd7eae8dc881fc5063edad06925d)"
  29440 Mar 15 11:16:09 V2R1C00B013-GUESTOS-FS-KVM-X64 systemd: Removed slice system-systemd\x2dfsck.slice.
  29441 Mar 15 11:16:09 V2R1C00B013-GUESTOS-FS-KVM-X64 sshd[1316]: Received signal 15; terminating.
  29442 Mar 15 11:16:09 V2R1C00B013-GUESTOS-FS-KVM-X64 systemd: Stopping system-systemd\x2dfsck.slice.
  29443 Mar 15 11:16:09 V2R1C00B013-GUESTOS-FS-KVM-X64 systemd: Stopped target Timers.
```
after system startup:
```
31700 Mar 15 11:17:25 V2R1C00B013-GUESTOS-FS-KVM-X64 docker: time="2017-03-15T11:17:25.130551833Z" level=debug msg="devmapper: constructDeviceIDMap() END"
  31701 Mar 15 11:17:25 V2R1C00B013-GUESTOS-FS-KVM-X64 docker: time="2017-03-15T11:17:25.138803387Z" level=debug msg="devmapper: Rolling back open transaction: TransactionID=149470 hash=0f1759dafc79227fabed4d2f2af8aa12963bdd7eae8dc881fc5        063edad06925d device_id=74887"
```

rhvgoyal commented:
@moypray I don't think sync is a problem. But one thing is odd.

Before crash we see the log "Unmount Device()", that means somebody has done Put() on device. That means creation of device must have been successful and there should not be any need to rollback.

But same device is rolled back. So I am wondering what happened.

BTW, how does your thin pool look like. Are you using loop devices for thin pool by any chance? If yes, dont use loop devices. Use proper block devices and retest.

Otherwise, paste more logs before crash. And first prove that device being rolled back had been created successfully and should not have been rolled back.

Once that is proven, put some logs in  processPendingTransaction() and figure out why rollback is happening. Ideally if transaction completed, then we update the thin pool transaction ID. Other possibility is that somehow that transaction ID did not make it to disk and that would be a bug. 

So put some more debug logs and try to narrow down it a bit.

moypray commented:
> But same device is rolled back. So I am wondering what happened.

I use `direct lvm+devicemapper` rather than `loop device`, I think the rollback operation is not related with the thin pool but the metadata on disk. So I still believe that the device metadata is not flushed to disk only in page cache before system restarts. The transaction and device metadata is not synced after os.Rename(), Why do you so sure about it?

Sorry the environment is recovered and thin pool is recreated. I just find the issue logs here. 
I will try to fallback my code, and reproduce it, but today I am so busy will attach here later. thanks

moypray commented:
> But same device is rolled back. So I am wondering what happened.

I use `direct lvm+devicemapper` rather than `loop device`, I think the rollback operation is not related with the thin pool but the metadata on disk. So I still believe that the device metadata is not flushed to disk only in page cache before system restarts. The transaction and device metadata is not synced after os.Rename(), Why do you so sure about it?

Sorry the environment is recovered and thin pool is recreated. I just find the issue logs here. 
I will try to fallback my code, and reproduce it, but today I am so busy and will attach here later. thanks

rhvgoyal commented:
- We do write new metadata to a tempory file, sync it to disk and then rename it. So I don't think flushing metadata to disk is a problem. check `saveTransactionMetaData`

- rollback is a combination of transaction id in transaction-metadata file and transaction id read from thin pool. If transaction id in transaction-metadata file is higher than in thin pool, then we know we started some transaction but it did not make into pool so we delete the device.


        // If there was open transaction but pool transaction ID is same
        // as open transaction ID, nothing to roll back.
        if devices.TransactionID == devices.OpenTransactionID {
                return nil
        }

LK4D4 commented:
@moypray are you still interested in working on this?

moypray commented:
Hi @LK4D4 @rhvgoyal 
Actually I have no idea how to fix it. I still want to fix it by an good way.

My fix is just a workaround about it , I don't know how does it happen.

moypray commented:
@LK4D4  I have my fix patched in my local version. And do not meet this issue again. If you think this is not an issue, you could close this issue. 

If I met this again, I will submit more logs here. thanks

rhvgoyal commented:
@moypray having a fix in your local docker is not enough justification to get something included upstream. We will need proper analysis of the issue along with patch so that people can understand what's the real issue and how it is being fixed.

moypray commented:
@rhvgoyal Agreed.  But I am not working on this issue for long time.

And I remember that we cherry-pick lots of sync-write metadata for layer/image operation PR and dm PRs after this issue comes out. Maybe this issue is caused by other metadata not sync to disk while reboot the system?

But now I have no time to work on this issue, Maybe later If I have time, I will fallback my code and to check why the transaction data  leaves on disk.

thaJeztah commented:
Let me close this PR for now, but let us know if additional changes are needed.

Thanks for working on this!

moypray commented:
OK, If I meet it again, will submit more logs and detailed information about it.
